File: week1/credit.c
#import <stdio.h>
#import <cs50.h>
#import <math.h>

long cardNum;
int numLength;

int firstDigits(difference) {
    int x = cardNum/pow(10, numLength-difference);
    return x;
}

//Checks the validity of the card
int cardVal(int length, long num) {
    //makes the int into an array
    int arr[length];
    for(int i = 0; i < length; i++) {
        arr[i] = num % 10;
        num /= 10;
    }
    //Luhns Algorithm
    int total = 0;
    for(int i = 1; i < length; i += 2) {
        int numProduct = arr[i]*2;
        if(numProduct > 9) {
            total += numProduct % 10;
            numProduct /= 10;
        }
        total += numProduct;
    }
    for(int i = 0; i < length; i += 2) {
        total += arr[i];
    }
    for(int i = 0; i < length; i++){
    }
   return total % 10;
}



int main(void) {
    while(numLength != 13 && numLength != 15 && numLength != 16){
        numLength = 0;
        cardNum = get_long("What's your card number? ");
        long tempNum = cardNum;
        while(tempNum > 0) {
            tempNum /= 10;
            numLength++;
        }
    }
    
    int cardModulo = cardVal(numLength, cardNum);    
    switch(numLength) {
        case 13: //VISA
            if(firstDigits(1) != 4 || cardModulo != 0){
                printf("INVALID\n");
                break;
            }
            printf("VISA\n");
            break;
        case 15://AMEX
            if((firstDigits(2) != 34 && firstDigits(2) != 37) || cardModulo != 0) {
                printf("INVALID\n");
                break;
            }
            printf("AMEX\n");
            break;
        case 16://MASTERCARD AND VISA
           if(cardModulo == 0) {
               if(firstDigits(1) == 4) {                   
                   printf("VISA\n");
                   break;
               }
               else if(firstDigits(2) > 50 && firstDigits(2) < 56) {
                   printf("MASTERCARD\n");
                   break;
               }
           }
            printf("INVALID\n");
            break;
    }
}

File: week6/credit.py
from cs50 import get_string


#Luhn's algorithm
def luhn(cardNum, company):
    cardLength = len(cardNum) - 2
    tempVal = 0
    tempCardNum = list(cardNum)
    while cardLength >= 0:
        tempCardNum[cardLength] = ''
        product = int(cardNum[cardLength]) * 2
        if product >= 10:
            tempVal += product % 10 + 1
        else:
            tempVal += product
        cardLength -= 2
        
    tempCardNum = ''.join(tempCardNum)
    for j in range(len(tempCardNum)):
        tempVal += int(tempCardNum[j])

    if tempVal % 10 == 0:
        print(company)
        return True
    
    print("INVALID")
    return False
        
        
#Checks size/card type and applies luhn's algorithm to it
def validation(cardNum):
    #VISA CASE
    if int(cardNum[0]) == 4 and (len(cardNum) == 13 or len(cardNum) == 16):
        luhn(cardNum, "VISA")
    
    #MASTERCARD CASE
    elif cardNum[0:2] in {"51", "52", "53", "54", "55"} and len(cardNum) == 16:
        luhn(cardNum,"MASTERCARD")
        
    #AMEX CASE
    elif cardNum[0:2] in {"34", "37"} and len(cardNum) == 15:
        luhn(cardNum, "AMEX")
    
    else:
        print("INVALID")
        return False
        

def main():
    while True:
        cardNum = get_string("Number: ")
        if cardNum.isdigit():
            break
    validation(cardNum)
    
    
    
main()